nifi-nar-bundles/nifi-cdc/nifi-cdc-mysql-bundle/nifi-cdc-mysql-processors/src/main/java/org/apache/nifi/cdc/mysql/processors/CaptureChangeMySQL.java

```diff
@@ -0,0 +1,1007 @@
+                lastStateUpdate = now;
+            }
+        } catch (IOException ioe) {
+            try {
+                // Perform some processor-level "rollback", then rollback the session
+                currentBinlogFile = xactBinlogFile == null ? "" : xactBinlogFile;
+                currentBinlogPosition = xactBinlogPosition;
+                currentSequenceId.set(xactSequenceId);
+                inTransaction = false;
+                stop(stateManager);
+                queue.clear();
+                currentSession.rollback();
+            } catch (Exception e) {
+                // Not much we can recover from here
+                log.warn("Error occurred during rollback", e);
+            }
+            throw new ProcessException(ioe);
+        }
+    }
+
+    @OnStopped
+    public void onStopped(ProcessContext context) {
+        try {
+            stop(context.getStateManager());
+        } catch (CDCException ioe) {
+            throw new ProcessException(ioe);
+        }
+    }
+
+    @OnShutdown
+    public void onShutdown(ProcessContext context) {
+        try {
+            // In case we get shutdown while still running, save off the current state, disconnect, and shut down gracefully
+            stop(context.getStateManager());
+        } catch (CDCException ioe) {
+            throw new ProcessException(ioe);
+        }
+    }
+
+    /**
+     * Get a list of hosts from a NiFi property, e.g.
+     *
+     * @param hostsString A comma-separated list of hosts (host:port,host2:port2, etc.)
+     * @return List of InetSocketAddresses for the hosts
+     */
+    private List<InetSocketAddress> getHosts(String hostsString) {
+
+        if (hostsString == null) {
+            return null;
+        }
+        final List<String> hostsSplit = Arrays.asList(hostsString.split(","));
+        List<InetSocketAddress> hostsList = new ArrayList<>();
+
+        for (String item : hostsSplit) {
+            String[] addresses = item.split(":");
+            if (addresses.length != 2) {
+                throw new ArrayIndexOutOfBoundsException("Not in host:port format");
+            }
+
+            hostsList.add(new InetSocketAddress(addresses[0].trim(), Integer.parseInt(addresses[1].trim())));
+        }
+        return hostsList;
+    }
+
+    protected void connect(List<InetSocketAddress> hosts, String username, String password, boolean createEnrichmentConnection,
+                           String driverLocation, String driverName, long connectTimeout) throws IOException {
+
+        int connectionAttempts = 0;
+        final int numHosts = hosts.size();
+        InetSocketAddress connectedHost = null;
+
+        while (connectedHost == null && connectionAttempts < numHosts) {
+            if (binlogClient == null) {
+
+                connectedHost = hosts.get(currentHost);
+                binlogClient = createBinlogClient(connectedHost.getHostString(), connectedHost.getPort(), username, password);
```

Comment:
No, I think it has to be unique. If MySQL cluster has three node (1, 2, 3) then NiFi processor (BinLogClient) can't have neither 1, 2 nor 3, it should have other ID. The library uses the possible maximum server Id ([65535](https://github.com/shyiko/mysql-binlog-connector-java/blob/master/src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java#L127)) as default value. Any binlog client will be a part of the cluster as a slave, it seems.